File: ios/SignLanguageTranslation/SignLanguageModule.swift
import Foundation
import UIKit
import React

/**
 The whole native surface of the SDK on iOS.

 v2 moved the player, the translation flow, the API client and tap
 classification into TypeScript, shared with Android. What is left is the one
 thing JavaScript cannot do: add a "Sign Language" item to the system's
 text-selection menu, for text the host app already made selectable.

 Three things the v1 module did are deliberately gone:

  - The 0.5 s `Timer` that walked every window's entire view hierarchy, attaching
    gesture recognisers, and which was never invalidated — not even by `disable()`.
  - The per-view tap and long-press recognisers. Tap-to-translate is now
    classified in JavaScript against the React tree, which is cheaper and can
    tell a button's label from its box.
  - The bottom sheet and its `AVPlayer`.

 What remains hooks exactly one class — `UITextView`, which is what backs React
 Native's selectable text on iOS — and only to offer a menu item.
 */
@objc(SignLanguageTranslation)
class SignLanguageModule: RCTEventEmitter {

    fileprivate static var isEnabled = false
    fileprivate static var language = "tr"
    fileprivate static weak var current: SignLanguageModule?

    private var hasListeners = false

    override init() {
        super.init()
        SignLanguageModule.current = self
    }

    // MARK: - RCTEventEmitter

    override static func requiresMainQueueSetup() -> Bool { true }

    override func supportedEvents() -> [String] { ["onTextSelected"] }

    override func startObserving() { hasListeners = true }
    override func stopObserving() { hasListeners = false }

    // MARK: - Bridge surface

    /// Only the language matters here: it localizes the menu item's title.
    @objc(configure:)
    func configure(_ language: String) {
        SignLanguageModule.language = language
        DispatchQueue.main.async { SelectionMenu.install() }
    }

    @objc(setEnabled:)
    func setEnabled(_ enabled: Bool) {
        SignLanguageModule.isEnabled = enabled
        DispatchQueue.main.async { SelectionMenu.install() }
    }

    /// Kept for API compatibility. Selection handling is global on iOS, so a
    /// per-view opt-in has nothing to do.
    @objc(enableTextSelectionForView:)
    func enableTextSelectionForView(_ viewTag: NSNumber) {}

    /**
     The key window's safe-area insets, in points.

     The SDK's floating surfaces keep themselves inside the reachable area, and
     the notch, the Dynamic Island and the home indicator are what define it.
     JavaScript can only guess at those from the screen size; the window knows.
     Resolves zeroes when there is no window yet, and the caller falls back to
     its own estimate.
     */
    @objc(getSafeAreaInsets:rejecter:)
    func getSafeAreaInsets(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        DispatchQueue.main.async {
            guard let window = Self.keyWindow else {
                resolve(nil)
                return
            }
            let insets = window.safeAreaInsets
            resolve([
                "top": insets.top,
                "bottom": insets.bottom,
                "left": insets.left,
                "right": insets.right,
            ])
        }
    }

    private static var keyWindow: UIWindow? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }
    }

    // MARK: - Selection

    static var menuTitle: String {
        switch language {
        case "ar": return "لغة الإشارة"
        case "tr": return "İşaret Dili"
        default: return "Sign Language"
        }
    }

    static var enabled: Bool { isEnabled }

    /// Called by the swizzled action when the user picks the menu item.
    static func report(text: String) {
        guard isEnabled,
              !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              let module = current,
              module.hasListeners
        else { return }

        module.sendEvent(withName: "onTextSelected", body: ["text": text])
    }
}

// MARK: - The menu item

/**
 Adds the item to the edit menu, and nothing else.

 The SDK must not force text to be selectable: v1 wrapped the app in a selection
 layer, which broke normal interaction. Selection-based translation is offered
 only where the host already made text selectable — and on iOS that is exactly
 where a `UITextView` exists.
 */
enum SelectionMenu {

    private static var installed = false

    static func install() {
        registerMenuItem()

        guard !installed else { return }
        installed = true

        // `UITextView` implements `canPerformAction(_:withSender:)` itself, so
        // exchanging here swaps that class's own implementation and leaves
        // every other responder alone.
        guard
            let original = class_getInstanceMethod(
                UITextView.self, #selector(UIResponder.canPerformAction(_:withSender:))),
            let replacement = class_getInstanceMethod(
                UITextView.self, #selector(UITextView.sl_canPerformAction(_:withSender:)))
        else { return }

        method_exchangeImplementations(original, replacement)
    }

    /// The item itself. Re-registered on each call so a language change lands.
    private static func registerMenuItem() {
        let item = UIMenuItem(
            title: SignLanguageModule.menuTitle,
            action: #selector(UITextView.sl_translate(_:))
        )
        let others = (UIMenuController.shared.menuItems ?? []).filter {
            $0.action != #selector(UITextView.sl_translate(_:))
        }
        UIMenuController.shared.menuItems = others + [item]
    }
}

extension UITextView {

    /// The action the menu item invokes.
    @objc func sl_translate(_ sender: Any?) {
        guard let text = sl_selectedText(), !text.isEmpty else { return }
        UIMenuController.shared.hideMenu()
        SignLanguageModule.report(text: text)
    }

    /// Swapped in for `canPerformAction`, to offer the item.
    ///
    /// The exchange means the call below reaches the original implementation,
    /// so every other menu item behaves exactly as it did.
    @objc func sl_canPerformAction(_ action: Selector, withSender sender: Any?) -> Bool {
        if action == #selector(UITextView.sl_translate(_:)) {
            return SignLanguageModule.enabled && !(sl_selectedText() ?? "").isEmpty
        }
        return sl_canPerformAction(action, withSender: sender)
    }

    /// The current selection, if there is one.
    fileprivate func sl_selectedText() -> String? {
        guard let range = selectedTextRange, !range.isEmpty else { return nil }
        return text(in: range)
    }
}

File: android/src/main/java/com/signlanguagetranslation/SignLanguageModule.kt
package com.signlanguagetranslation

import android.graphics.Point
import android.os.Build
import android.view.View
import android.view.ViewGroup
import android.view.WindowInsets
import android.widget.TextView
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.Arguments
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.common.ViewUtil
import com.signlanguagetranslation.textselection.CustomActionModeCallback

/**
 * The whole native surface of the SDK on Android.
 *
 * v2 moved the player, the translation flow, the API client and tap
 * classification into TypeScript, shared with iOS. What is left here is the one
 * thing JavaScript genuinely cannot do: add a "Sign Language" item to the
 * platform's text-selection action mode.
 *
 * Two things the v1 module did are deliberately gone:
 *
 *  - The `Handler` loop that re-walked the entire view hierarchy every 2000 ms,
 *    reinstalling callbacks on every pass and never stopping, even after
 *    `disable()`.
 *  - The per-view tap `GestureDetector`. Tap-to-translate is now classified in
 *    JavaScript against the React tree, which is both cheaper and able to tell
 *    a button's label from its box.
 */
class SignLanguageModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var enabled = false
    private var language = "tr"
    private var listenerCount = 0

    override fun getName(): String = NAME

    /** Only the language matters here: it localizes the menu item's title. */
    @ReactMethod
    fun configure(language: String) {
        this.language = language
    }

    @ReactMethod
    fun setEnabled(enabled: Boolean) {
        this.enabled = enabled
        if (enabled) {
            reactApplicationContext.runOnUiQueueThread { attachToVisibleText() }
        }
    }

    /**
     * Opt one view subtree into the selection menu.
     *
     * Hosts that render text through a custom view can call this; the common
     * case is handled by [setEnabled], which sweeps the current screen once.
     */
    @ReactMethod
    fun enableTextSelectionForView(viewTag: Int) {
        reactApplicationContext.runOnUiQueueThread {
            // Resolved through UIManagerHelper rather than UIManagerModule:
            // the latter exists only on the old architecture, so on the new one
            // it returns null and this would silently do nothing.
            val view = try {
                UIManagerHelper
                    .getUIManager(reactApplicationContext, ViewUtil.getUIManagerType(viewTag))
                    ?.resolveView(viewTag)
            } catch (error: Exception) {
                null
            }
            if (view != null) attach(view)
        }
    }

    /**
     * The system-bar insets that actually cover the app's content, in dp.
     *
     * The SDK's floating surfaces keep themselves inside the reachable area,
     * and on Android that area cannot be guessed: since targetSdk 35 the window
     * is edge-to-edge by default, so the navigation bar overlaps the content
     * the app lays out, while an older or opted-out app gets a window that
     * already stops above it. Guessing either way is wrong on the other.
     *
     * So this reports the *overlap*, not the raw inset: how far each system bar
     * reaches into the content view, which is zero when the window already
     * excludes it. Resolves `null` when there is no attached window to measure,
     * and the caller falls back to its own estimate.
     */
    @ReactMethod
    fun getSafeAreaInsets(promise: Promise) {
        reactApplicationContext.runOnUiQueueThread {
            promise.resolve(
                try {
                    contentInsets()
                } catch (error: Exception) {
                    null
                }
            )
        }
    }

    private fun contentInsets(): WritableMap? {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) return null

        val activity = reactApplicationContext.currentActivity ?: return null
        // The decor view, not the content view: React Native's window
        // dimensions describe the *window*, and an inset only means something
        // measured in the same space as the layout it is applied to. The
        // content view can be inset by a further amount the app never sees,
        // which would leave the reported overlap short of the real bar.
        val decor = activity.window?.decorView ?: return null
        val windowInsets = decor.rootWindowInsets ?: return null

        val barLeft: Int
        val barTop: Int
        val barRight: Int
        val barBottom: Int
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            // The cutout counts too: a landscape notch eats into the same edge
            // a card would otherwise be flush against.
            val bars = windowInsets.getInsets(
                WindowInsets.Type.systemBars() or WindowInsets.Type.displayCutout()
            )
            barLeft = bars.left
            barTop = bars.top
            barRight = bars.right
            barBottom = bars.bottom
        } else {
            @Suppress("DEPRECATION")
            barLeft = windowInsets.systemWindowInsetLeft
            @Suppress("DEPRECATION")
            barTop = windowInsets.systemWindowInsetTop
            @Suppress("DEPRECATION")
            barRight = windowInsets.systemWindowInsetRight
            @Suppress("DEPRECATION")
            barBottom = windowInsets.systemWindowInsetBottom
        }

        val displayWidth: Int
        val displayHeight: Int
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val bounds = activity.windowManager.currentWindowMetrics.bounds
            displayWidth = bounds.width()
            displayHeight = bounds.height()
        } else {
            val size = Point()
            @Suppress("DEPRECATION")
            activity.windowManager.defaultDisplay.getRealSize(size)
            displayWidth = size.x
            displayHeight = size.y
        }

        // Where the window sits on the display. Whatever lies between its edges
        // and the display's is already outside the app's layout, and pays for
        // that much of the bar — which is exactly what a window that is not
        // edge-to-edge does, and why its overlap comes out zero.
        val location = IntArray(2)
        decor.getLocationOnScreen(location)
        val outsideLeft = location[0]
        val outsideTop = location[1]
        val outsideRight = displayWidth - (location[0] + decor.width)
        val outsideBottom = displayHeight - (location[1] + decor.height)

        val density = decor.resources.displayMetrics.density
        fun overlap(bar: Int, outside: Int): Double =
            (maxOf(0, bar - maxOf(0, outside)) / density).toDouble()

        return Arguments.createMap().apply {
            putDouble("top", overlap(barTop, outsideTop))
            putDouble("bottom", overlap(barBottom, outsideBottom))
            putDouble("left", overlap(barLeft, outsideLeft))
            putDouble("right", overlap(barRight, outsideRight))
        }
    }

    /** Required by NativeEventEmitter. */
    @ReactMethod
    fun addListener(eventType: String) {
        listenerCount += 1
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        listenerCount = maxOf(0, listenerCount - count)
    }

    private fun attachToVisibleText() {
        // Called explicitly rather than as `currentActivity`: React Native
        // converted this class to Kotlin, so the Java synthetic property no
        // longer exists, and the inherited accessor is deprecated in favour of
        // going through the context.
        val activity = reactApplicationContext.currentActivity ?: return
        val root = activity.findViewById<View>(android.R.id.content) ?: return
        attach(root)
    }

    /**
     * Install the action-mode callback on every [TextView] in a subtree.
     *
     * Note this does **not** make text selectable. v1 wrapped the app in a
     * selection layer, which broke normal interaction; selection-based
     * translation is offered only where the host already made text selectable.
     */
    private fun attach(view: View) {
        if (view is TextView) {
            if (view.isTextSelectable) {
                view.customSelectionActionModeCallback =
                    CustomActionModeCallback(view, menuTitle()) { text -> emitSelection(text) }
            }
            return
        }
        if (view is ViewGroup) {
            for (index in 0 until view.childCount) attach(view.getChildAt(index))
        }
    }

    private fun menuTitle(): String = when (language) {
        "ar" -> "لغة الإشارة"
        "tr" -> "İşaret Dili"
        else -> "Sign Language"
    }

    private fun emitSelection(text: String) {
        if (!enabled || text.isBlank() || listenerCount == 0) return

        val payload: WritableMap = Arguments.createMap().apply { putString("text", text) }
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit("onTextSelected", payload)
    }

    companion object {
        const val NAME = "SignLanguageTranslation"
    }
}
